### PR TITLE
feat(client-interaction): Supporting idle option

### PR DESCRIPTION
**Pull request by leomp12**
Use `client:interaction="idle"` for first interaction -> idle callback -> hydrate

### CODE REVIEW

File: packages/client-interaction/runtime/directive.ts [hunk 1]
```diff
@@ -9,8 +9,20 @@ const firstInteraction = new Promise<void>(resolve => {
...
+    const hy = async () => {
+        const hydrate = await load()
+        await hydrate()
+    }
+    const arrOpts = Array.isArray(opts.value) ? opts.value : [opts.value]
+    if (arrOpts.includes('idle')) {
+        if (typeof window.requestIdleCallback === 'function') {
+            window.requestIdleCallback(hy)
+            return
+        }
```
Comment (lilnasy) [inline packages/client-interaction/runtime/directive.ts:23]: If I understand correctly, the idea is to delay the loading the page is also idle. Could you elaborate on when that would be preferable?

**Review by lilnasy — COMMENTED** [review]
This looks great, thanks for implementing! I have a question about when someone might use this.

File: packages/client-interaction/runtime/directive.ts (hunk 1 above)
Comment (leomp12) [reply packages/client-interaction/runtime/directive.ts:23]: Hey, sorry for the lack of explanation 🙃

The idle option is to wait the next browser idle period to call component hydration.
When having many hydrated components you probably have some `client:idle` and probably believe it'll be hydrated before `client:interaction` (and after `client:load`), but it's not necessarily true cause first interaction may occur before first idle.

Also, if using view transitions `client:interaction` will act like `client:load` on navigation, sometimes even before `client:load`, in concurrency with (and thus delaying) more important components.

Comment (lilnasy) [reply packages/client-interaction/runtime/directive.ts:23]: Thanks, that makes a lot of sense!

The big consideration now is the size of the directive. For performance reasons, the code for the directive is inlined into html, which means if there are 5 islands using `client:interaction`, there will be 5 copies of this code in the initial response.

Do you think it makes sense for it to be another directive (with fewer lines of code) instead of an option for `client:interaction`?

Comment (leomp12) [reply packages/client-interaction/runtime/directive.ts:23]: Are you sure directive code is replicated each island? 🤔 
I didn't know that, if it is true I will have to review a custom directive that I use, but from what I'm testing the code only appears once even with more than one component with the directive on the page

Comment (leomp12) [reply packages/client-interaction/runtime/directive.ts:23]: Take a look https://demo.ecomplus.app/p-blusa-charlotte-ml-cinza-mid
In the output HTML code we have two `client="context"`, but only one `addEventListener("storefront:apiContext"`

Comment (leomp12) [reply packages/client-interaction/runtime/directive.ts:23]: An I missing something?

Comment (lilnasy) [reply packages/client-interaction/runtime/directive.ts:23]: You're right! It is inlined but included at most once per page.

File: packages/client-interaction/runtime/directive.ts [hunk 2]
```diff
@@ -9,8 +9,20 @@ const firstInteraction = new Promise<void>(resolve => {
...
-export default (async load => {
+export default (async (load, opts) => {
     await firstInteraction
-    const hydrate = await load()
-    await hydrate()
+    const hy = async () => {
+        const hydrate = await load()
+        await hydrate()
+    }
+    const arrOpts = Array.isArray(opts.value) ? opts.value : [opts.value]
```
Comment (lilnasy) [inline packages/client-interaction/runtime/directive.ts:18]: `opts.value` being an array does not align with its typescript type. Why was this necessary?

Comment (leomp12) [reply packages/client-interaction/runtime/directive.ts:18]: Options can be set like `client:context={['data:categories','data:grids']}`, but anyway here it would not be necessary, maybe just to avoid the need for a breaking change in a future with another option wanted for this `client:interaction` directive

Comment (lilnasy) [reply packages/client-interaction/runtime/directive.ts:18]: I don't think a breaking change will be necessary. We can introduce another value type anytime, like we are already doing by allowing a string now.

Comment (leomp12) [reply packages/client-interaction/runtime/directive.ts:18]: You're right, my bad...

**Review by lilnasy — APPROVED** [review]
Looks great to me! Again thanks for taking this on!

Would like a confirmation that the communication for the feature includes everything you would expect, and misses nothing, then it's ready to release.

**Comment by leomp12** [comment]
Communication is perfect IMO 👍🏾

**Comment by lilnasy** [comment]
Awesome, let's ship it! 🚀
